scipy/spatial/_lloyd_algorithm.py

```diff
@@ -0,0 +1,221 @@
+
+import numpy as np
+from typing import (
+    Optional,
+    TYPE_CHECKING,
+)
+
+if TYPE_CHECKING:
+    import numpy.typing as npt
+    from scipy._lib._util import (
+        DecimalNumber, IntNumber
+    )
+
+from scipy.spatial import distance, Voronoi
+
+
+__all__ = ['lloyd_centroidal_voronoi_tessellation']
+
+
+def _l1_norm(points: np.ndarray) -> float:
+    l1 = distance.cdist(points, points, 'cityblock')
+    return np.min(l1[l1.nonzero()])
+
+
+def _lloyd_centroidal_voronoi_tessellation(
+        points: np.ndarray,
+        decay: float,
+        qhull_options: str,
+) -> np.ndarray:
+    """Lloyd-Max algorithm iteration.
+
+    Based on the implementation of Stéfan van der Walt:
+
+    https://github.com/stefanv/lloyd
+
+    which is:
+
+        Copyright (c) 2021-04-21 Stéfan van der Walt
+        https://github.com/stefanv/lloyd
+        MIT License
+
+    Parameters
+    ----------
+    points : array_like (n, d)
+        The points to iterate on.
+    decay : float
+        Relaxation decay. A positive value would move the points toward
+        their centroid, and negative value would move them away.
+        1 would move the points to their centroid.
+    qhull_options : str
+        Additional options to pass to Qhull. See Qhull manual
+        for details. (Default: "Qbb Qc Qz Qj Qx" for ndim > 4 and
+        "Qbb Qc Qz Qj" otherwise.)
+
+    Returns
+    -------
+    points : array_like (n, d)
+        The points after an iteration of Lloyd's algorithm.
+
+    """
+    new_points = np.empty_like(points)
+
+    voronoi = Voronoi(points, qhull_options=qhull_options)
+
+    for ii, idx in enumerate(voronoi.point_region):
+        # the region is a series of indices into self.voronoi.vertices
+        # remove point at infinity, designated by index -1
+        region = [i for i in voronoi.regions[idx] if i != -1]
+
+        # get the vertices for this region
+        verts = voronoi.vertices[region]
+
+        # clipping would be wrong, we need to intersect
+        # verts = np.clip(verts, 0, 1)
+
+        # move points towards centroids:
+        # Centroid in n-D is the mean for uniformly distributed nodes
+        # of a geometry.
+        centroid = np.mean(verts, axis=0)
+        new_points[ii] = points[ii] + (centroid - points[ii]) * decay
+
+    # only update points to centroid within the region
+    is_valid = np.all(np.logical_and(new_points >= 0, new_points <= 1), axis=1)
+    points[is_valid] = new_points[is_valid]
+
+    return points
+
+
+def lloyd_centroidal_voronoi_tessellation(
+        points: npt.ArrayLike,
+        *,
+        tol: DecimalNumber = 1e-5,
+        maxiter: IntNumber = 10,
+        decay: Optional[npt.ArrayLike] = None,
+        qhull_options: Optional[str] = None,
+) -> np.ndarray:
+    """Approximate Centroidal Voronoi Tessellation.
+
+    Perturb points in :math:`[0, 1]^d` using Lloyd-Max algorithm.
+
+    Parameters
+    ----------
+    points : array_like (n, d)
```

Comment:
I don't think we should be mutating the input data structure--when I run locally, instead of returning a new set of points, the original data structure gets mutated, which is not something I'd expect with most SciPy functions, at least not in `spatial`.

Did @mdhaber ask for that or?

```python
import numpy as np
import scipy
from scipy.spatial import lloyd_centroidal_voronoi_tessellation

orig_points = np.array([[0.1, 0.1],
                        [0.1, 0.2],
                        [0.2, 0.1],
                        [0.2, 0.2]])
new_points = lloyd_centroidal_voronoi_tessellation(points=orig_points)
print("orig_points:", orig_points)
#orig_points: [[0.14999843 0.14999843]
# [0.14999843 0.15000157]
# [0.15000157 0.14999843]
# [0.15000157 0.15000157]]
```

Reply:
And just to be clear--even if folks do decide that it is "ok" to mutate the input data structure here, I'd certainly be surprised if folks were in favor of a function that *both* mutates its input *and* returns the mutated input, since not returning anything would make the mutating nature of the function clearer.

Reply:
Ah right, nothing was asked for that. Would `points = np.asarray(points).copy()` be fine?

Reply:
Seems to work--the regression test below fails before/passes after, so perhaps worth adding that as well to lock it down moving forward. You may notice that one of the other tests start to fail with the `copy()` added in, but that seems like useful information anyway.

```diff
--- a/scipy/spatial/_lloyd_algorithm.py
+++ b/scipy/spatial/_lloyd_algorithm.py
@@ -177,7 +177,7 @@ def lloyd_centroidal_voronoi_tessellation(
     0.0278...  # random
 
     """
-    points = np.asarray(points)
+    points = np.asarray(points).copy()
 
     if not points.ndim == 2:
         raise ValueError('Sample is not a 2D array')
diff --git a/scipy/spatial/tests/test_lloyd_algorithm.py b/scipy/spatial/tests/test_lloyd_algorithm.py
index 08f2551d3..c9807e2c0 100644
--- a/scipy/spatial/tests/test_lloyd_algorithm.py
+++ b/scipy/spatial/tests/test_lloyd_algorithm.py
@@ -67,3 +67,17 @@ def test_lloyd_errors():
         lloyd_centroidal_voronoi_tessellation(
                 points, decay=decay, maxiter=maxiter
         )
+
+
+def test_lloyd_non_mutating():
+    # verify that the input points are not
+    # mutated in place and that they do not
+    # share memory with the output
+    orig_points = np.array([[0.1, 0.1],
+                            [0.1, 0.2],
+                            [0.2, 0.1],
+                            [0.2, 0.2]])
+    points_copy = orig_points.copy()
+    new_points = lloyd_centroidal_voronoi_tessellation(points=orig_points)
+    assert_allclose(orig_points, points_copy)
+    assert not np.may_share_memory(orig_points, new_points)
```

Other fail I see locally after adding the `copy()`:

```
scipy/spatial/tests/test_lloyd_algorithm.py:33: in test_lloyd
    assert base_l1 < curr_l1
E   assert 0.01792591639156088 < 0.01792591639156088
```

Reply:
Great thanks I will add this! For the failure this one is not very robust. I will have a look.